### ISSUE TITLE
Commentary on a pure Redux approach

### ISSUE THREAD

**Issue by bodhi** [slorber/scalable-frontend-with-elm-or-redux#24]
I'm just finishing up a (semi-cheating) attempt at this, and when writing it up I also wrote up some notes on doing it purely inside Redux. I figured this might be worth sharing as well, but since it's not specific to my implementation I didn't want to put it in the Readme.

Thanks @slorber by the way for formulating this, it's quite a useful thought exercise!

---

For a pure-Redux implementation, the individual parts of each system are quite simple, the difficulty manifests in putting the app together: The counter needs the _state_ from a separate part of the app (the button), combined with _the dispatched actions_ from multiple _unknown_ parts (all of the NewGifs).

---

There are (at least) a few ways to share the button _state_ with the counter:
1. Have the counter reducer listen for the action that the button triggers (eg. `SET_BUTTON` or `TOGGLE_BUTTON`). This couples the button and the counter on the action.
2. Put the counter reducer at a higher level (in a `combineReducers`-sense) in the reducer tree than the button reducer, so that it can see the button's portion of the state. This couples the counter and the button on the button's state definition.
3. Have a reducer that wraps the counter reducer that maps a `NEW_GIF` action to a new action that includes the state of the button in the action payload. (... What's the aphorism about fixing everything by adding an abstraction layer?)

---

The real difficulty though is allowing the counter to see _every_ `NEW_GIF` action, while restricting each `NewGif` to only see/process actions that are _relevant to it_ (to avoid every `NewGif` instance switching to the same latest gif).

You need to start namespacing them in some way, and the counter becomes coupled to whatever namespacing scheme you cook up...

---

Basically, counter is screaming out for some kind of event bus, but the only event bus available in plain Redux is the reducers.

**Comment by slorber** [comment]
@bodhi that's why I proposed somehow 2 mailboxes.

> You need to start namespacing them in some way, and the counter becomes coupled to whatever namespacing scheme you cook up...

it does not look to me a problem if this is an event-bus of global business events. In this app, NewGif is clearly a business event on which you build business rules. The counter on its own does not have to be coupled but you could provide an integration layer (wrapper) on top of it that couples the reusable component to your app and your business events bus.

**Comment by bodhi** [comment]
> it does not look to me a problem if this is an event-bus of global business events.

Definitely, my only point was that this concept doesn't really exist in plain Redux (or rather, no one really discusses doing it).

It really boils down to how the 4 teams agree to communicate. And I guess this repo is all about demonstrations/prototypes of different communication strategies!

**Comment by slorber** [comment]
yes it is

I've thought a bit about doing it on top of Redux by using a channel system. I need that channel system for UI perf improvements in `connect` implementation but did not work on that yet
